### ISSUE TITLE
`PDBFixer.addSolvent` often incorrectly neutralises systems with charged ligands as though they were neutral

### ISSUE THREAD

**Issue by Yoshanuikabundi** [openmm/pdbfixer#299]
The `addSolvent` method uses a force field to assign charges, which it then uses to add ions that neutralise the net charge of the box. Unfortunately, the force field seems to assign formal charges of zero to many charged groups in small molecules; for example the citrate ion, which has a -3 formal charge:

```python
from openff.toolkit import Molecule

top = Molecule.from_smiles("C(C(=O)[O-])C(CC(=O)[O-])(C(=O)[O-])O").to_topology()
top.molecule(0).generate_conformers(n_conformers=1)
top.to_file("citrate.pdb")

from pdbfixer import PDBFixer
import openmm

fixer = PDBFixer("citrate.pdb")

fixer.addSolvent(
    padding=2.0 * openmm.unit.nanometer,
    ionicStrength=0.1 * openmm.unit.molar,
    boxShape="dodecahedron",
)

counts = {"NA": 0, "UNK": 0, "CL": 0, "HOH": 0}
for residue in fixer.topology.residues():
    counts[residue.name] += 1
counts # 
{'NA': 3, 'UNK': 1, 'CL': 3, 'HOH': 1457}
```

The `citrate.pdb` file generated above is available here: [citrate.pdb.zip](https://github.com/user-attachments/files/16259296/citrate.pdb.zip)

With the current `main` branch (commit 92044f3), you can also see this behavior with a PDB file like [1PO0](https://files.rcsb.org/download/1PO0.pdb), which has 2 citrate ions (each with -3 formal charge).

```python
from pdbfixer import PDBFixer
import openmm

fixer = PDBFixer("1PO0.pdb")

fixer.findMissingResidues()

# This section adds caps; leave it commented to rebuild terminal loops
chains_to_cap = {chain for chain, resi in fixer.missingResidues}
for chainidx in chains_to_cap:
    chain = [*fixer.topology.chains()][chainidx]
    last_resi = len([*chain.residues()])
    fixer.missingResidues[chainidx, 0] = ['ACE']
    fixer.missingResidues[chainidx, last_resi] = ['NME']

fixer.findNonstandardResidues()
fixer.replaceNonstandardResidues()
# fixer.removeHeterogens(keepWater=True)
fixer.findMissingAtoms()
fixer.addMissingAtoms()

fixer.addMissingHydrogens(pH=7.4)

fixer.addSolvent(
    padding=2.0 * openmm.unit.nanometer,
    ionicStrength=0.1 * openmm.unit.molar,
    boxShape="dodecahedron",
)

charges = {
 'ASP': -1,
 'ARG': +1,
 'GLU': -1,
 'LYS': +1,
 'FLC': -3,
 'NA': 1,
 'CL': -1
} # I think this is right!
charge = 0
for residue in fixer.topology.residues():
    if residue.name in charges:
        charge += charges[residue.name]
charge # gives -6 (-3 for each citrate residue)
```

It would be great if the correct formal charges could be stored when the template is downloaded! Then they could be used when neutralising the system. This might also give an opportunity for users to manually specify the formal charges of `UNK` residues.

Thanks!

(also the ACE and NME residues in the second example above don't seem to be getting hydrogens - let me know if you'd like me to open a second issue for that!)

**Comment by peastman** [comment]
That's an interesting corner case I hadn't thought about.  `PDBFixer.addSolvent()` is mostly just a thin wrapper around `Modeller.addSolvent()`:

https://github.com/openmm/pdbfixer/blob/92044f3c524d21ee1e79f53c54882201ff2b5b34/pdbfixer/pdbfixer.py#L1358-L1364

Notice that Modeller lets you pass in a force field to use for identifying charges, but PDBFixer generates its own force field, which of course only knows about standard residues.  Here are a few possibilities.

- As a short term workaround, you could just copy out the code from `PDBFixer.addSolvent()`, but pass your own force field instead of having it generate one.
- We could modify `PDBFixer.addSolvent()` to let you pass in a force field to use.
- We could try to make `_createForceField()` generate charges for nonstandard residues based on their formal charges.

A complication with the third one is that it often doesn't know what the formal charges are.  Registering a template is only required if you're going to add new atoms to a residue.  If all you're doing is preserving a molecule from the input file, it will happily copy it over even if it has no idea what that molecule is.  It doesn't need to know anything about it... except in this case, it kind of does because it affects how many ions get added.

**Comment by Yoshanuikabundi** [comment]
Thanks @peastman, that clarifies the issues really well for me! I might take a crack at option 3 in a fork - would you accept a PR if I can come up with a good solution?

**Comment by peastman** [comment]
If you can come up with a good solution, sure, but let's discuss it first.  What do you have in mind for figuring out the charges?

If we can't figure out anything robust, I'd vote for the second one.  It's easy to implement, and if you plan to simulate the system with OpenMM, you presumably have a force field that can parameterize it.

**Comment by j-wags** [comment]
I'd be happy to try using @Yoshanuikabundi's fork as a way to evaluate implementing Peter's option 3 via either taking formal charges from the CCD, or using the MDAnalysis chemical information guesser at scale. These would both depend on the user knowing their protonation state beforehand, so users could either use the already-implemented protonation functionality, or they could come with fully-protonated inputs. 

Unfortunately, Peter's options 1 and 2 would hit a circular dependency in SMIRNOFF force field land. We can't make a FF for a molecule without already knowing its formal charges.

**Comment by peastman** [comment]
I think that's orthogonal to this issue?  This is only about how to determine the total charge for purposes of adding counter ions.  Parameterizing the system is much more complicated.  And if you have a solution for parameterizing, then allowing a force field to be passed to `addSolvent()` should handle this case.

**Comment by Yoshanuikabundi** [comment]
Option 2 would be fine if we had an OpenMM force field that could charge, say, anything in the PDB... but unfortunately we don't. And as you point out, if we did have an appropriate force field we could just use Modeller to add solvent. The `createForceField` machinery in PDBFixer is part of what makes it such a flexible tool, which is why I think the best path forward is to improve its ability to assign a wide variety of correct formal charges (option 3). I'll also point out that PDBFixer is a tool that can be used for preparing systems for any MD engine, and it would be a shame to tie its full functionality to any specific force field format - in the current API, this relationship is entirely implicit.

I'd propose adding a method to the `PDBFixer` class that uses the existing machinery for collecting CCD templates to assign formal charges to all atoms. Formal charges could be stored as a `list[int|None]` on the `PDBFixer` object, or as a `dict[int, int]` if we expect formal charges to be sparsely populated and the size of the list to be an issue. This would also let users manually set formal charges for unknown ligands or alternate protonation states.

Then `createForceField` could use those formal charges when assigning its charges. This would avoid any unnecessary CCD lookup for users who do not require formal charges or solvation. When the CCD is looked up, formal charges could be set at that time to avoid a future duplicate lookup. `createForceField` could issue a warning if it has any ligands with unknown formal charges, or it could call the new API point to determine charges.

I think @j-wags point is that OpenFF does have a solution for parametrizing, but (a) it isn't in the form of an OpenMM force field, and (b) it requires formal charges.

I'm not familiar enough with the new code to know off the top of my head how much of a refactor the above is, but I'm happy to experiment in a fork! I'm happy to present a maximalist solution to see if you like it and then have it whittled down in code review.

**Comment by peastman** [comment]
This was fixed by #301.
